torch/lib/THD/base/data_channels/DataChannelNccl.cpp

```diff
@@ -0,0 +1,731 @@
+                               "collective operations");
+    }
+    // Check the tensor type is identical
+    if (input[i].type().scalarType() != elementType ||
+        output[i].type().scalarType() != elementType) {
+      throw std::runtime_error("Expecting all GPU tensors to have identical "
+                               "type");
+    }
+    // Check the input tensor size is identical
+    if (input[i].numel() != inputNumElement) {
+      throw std::runtime_error("Expecting all input tensors to have identical "
+                               "number of elements");
+    }
+    // Check the output tensor size equals to input tensor size
+    if (output[i].numel() != inputNumElement * outputOverInput) {
+      throw std::runtime_error("The number of elements of output tensor does "
+                               "not match the number of elements of the input "
+                               "tensor");
+    }
+    // Contiguous verification
+    if (!input[i].is_contiguous() || !output[i].is_contiguous()) {
+      throw std::runtime_error("Expecting all GPU tensors to be contiguous");
+    }
+
+    auto res = usedDevices.insert(input[i].get_device());
+    // Device verification, if the insertion didn't take place
+    if (!res.second) {
+      throw std::runtime_error("Expecting inputs on different GPU devices");
+    }
+
+    // Now check the output device
+    if (input[i].get_device() != output[i].get_device()) {
+      throw std::runtime_error("Expecting input and output tensors to be on "
+                               "the same device");
+    }
+  }
+  return true;
+}
+
+
+void DataChannelNccl::allReduce(std::vector<at::Tensor>& input,
+                                std::vector<at::Tensor>& output,
+                                THDReduceOp operation,
+                                THDGroup groupId) {
+
+  std::unique_lock<std::mutex> channelLock(_mutex);
+  // Check the tensor vector for consistency
+  if (!_tensorCheckHelper(input, output)) {
+    return;
+  }
+  _checkGroupIdValid(groupId);
+
+  auto commsAndEvents = _getNcclCommsAndEvents(input, groupId);
+  auto comms = commsAndEvents.first;
+  auto events = commsAndEvents.second;
+
+  // Guard GPU device
+  AutoGPU gpuGuard;
+
+  std::unique_lock<std::mutex> cudaFreeMutexLock(
+      *(THCCachingAllocator_getCudaFreeMutex()));
+
+  NCCL_CHECK(ncclGroupStart());
+  for (size_t i = 0; i < input.size(); ++i) {
+
+    gpuGuard.setDevice(input[i].get_device());
+    auto stream = THCState_getCurrentStream(THDGetCudaState());
+
+    NCCL_CHECK(ncclAllReduce(input[i].data_ptr(),
+                             output[i].data_ptr(),
+                             input[i].numel(),
+                             _getNcclDataType(input[i].type().scalarType()),
+                             ncclOp[operation],
+                             (*comms)[i],
+                             stream));
+    THCudaCheck(cudaEventRecord((*events)[i], stream));
```

Comment:
What do we use the events for?

Reply:
Now these events are still needed for destroyGroup and destroyDataChannel functions